### PR TITLE
OHRM5X-1660: Admin - LDAP Configuration Screen

### PR DESCRIPTION
**Pull request by Super-Chama**

### CODE REVIEW

**Comment by Super-Chama** [comment]
Keep draft until unit tests added for new rules

File: src/plugins/orangehrmAdminPlugin/Controller/LdapConfigurationController.php [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    {
+        $component = new Component('ldap-configuration');
+        $this->setComponent($component);
+    }
+
+    /**
+     * @param Request $request
+     * @return Response
+     */
+    public function testConnection(Request $request): Response
```
Comment (Super-Chama) [inline src/plugins/orangehrmAdminPlugin/Controller/LdapConfigurationController.php:43]: @SandeepaOHRM check this mock API structure for test connection.

File: src/plugins/orangehrmLDAPAuthenticationPlugin/Api/LDAPConfigAPI.php [hunk 2]
```diff
@@ -129,96 +118,98 @@ public function create(): EndpointResult
...
-                $this->getRequestParams()->getString(
-                    RequestParams::PARAM_TYPE_BODY,
-                    self::PARAMETER_DISTINGUISHED_PASSWORD
-                )
+
+            $password = $this->getRequestParams()->getStringOrNull(
+                RequestParams::PARAM_TYPE_BODY,
+                self::PARAMETER_DISTINGUISHED_PASSWORD
             );
+            if (!is_null($password)) {
```
Comment (Super-Chama) [inline src/plugins/orangehrmLDAPAuthenticationPlugin/Api/LDAPConfigAPI.php:148]: @SandeepaOHRM If password is set already and user update other fields the password is not sent from FE. here is prevent updating password is null is sent from FE

File: src/plugins/orangehrmLDAPAuthenticationPlugin/Api/Traits/LDAPDataMapParamRuleCollection.php [hunk 3]
```diff
@@ -67,7 +75,7 @@ protected function getParamRuleCollection(): ParamRuleCollection
             $this->getValidationDecorator()->notRequiredParamRule(
                 new ParamRule(
                     self::PARAMETER_EMPLOYEE_ID,
-                    new Rule(Rules::EMAIL),
+                    new Rule(Rules::STRING_TYPE),
```
Comment (Super-Chama) [inline src/plugins/orangehrmLDAPAuthenticationPlugin/Api/Traits/LDAPDataMapParamRuleCollection.php:78]: @SandeepaOHRM  This is not email validation. this is email attribute field so i changed it

**Review by RajithaKumara — APPROVED** [review]

File: src/client/src/core/util/validation/rules.ts [hunk 4]
```diff
@@ -636,3 +636,37 @@ export const validLangString = function(value: string) {
 export const validSelection = function(value: string | object | null) {
   return typeof value === 'string' ? translate('general.invalid') : true;
 };
+
+export const validHostnameFormat = function(value: string): boolean | string {
```
Comment (RajithaKumara) [inline src/client/src/core/util/validation/rules.ts:640]: Not sure whether you considered, [Internationalized domain name](https://en.wikipedia.org/wiki/Internationalized_domain_name), [Punycode](https://en.wikipedia.org/wiki/Punycode)

E.g.
https://github.com/orangehrm/orangehrm-os-mobile/blob/master/src/lib/helpers/url.ts#L27
https://github.com/symfony/symfony/blob/6.1/src/Symfony/Component/Validator/Constraints/UrlValidator.php#L24-L46

File: src/client/src/core/util/validation/rules.ts [hunk 5]
```diff
@@ -636,3 +636,37 @@ export const validLangString = function(value: string) {
...
+  let fqdnRegex;
+
+  // If first character is number, treat as ip address. else hostname
+  if (/^\d+$/.test(value.charAt(0))) {
+    fqdnRegex = /^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])$/;
```
Comment (RajithaKumara) [inline src/client/src/core/util/validation/rules.ts:645]: are we handling IPv4 and IPv6

File: src/plugins/orangehrmAdminPlugin/Controller/LdapConfigurationController.php (hunk 1 above)
Comment (RajithaKumara) [reply src/plugins/orangehrmAdminPlugin/Controller/LdapConfigurationController.php:49]: TODO:: remove

File: src/plugins/orangehrmAdminPlugin/config/routes.yaml [hunk 6]
```diff
@@ -603,3 +603,18 @@ admin_language_translation:
...
+
+admin_ldap_test_connection:
+  path: /api/v2/admin/ldap/connect
+  controller: OrangeHRM\Admin\Controller\LdapConfigurationController::testConnection
+  methods: [ POST ]
+
+admin_ldap_sync_connection:
+  path: /api/v2/admin/ldap/sync
+  controller: OrangeHRM\Admin\Controller\LdapConfigurationController::syncConnection
+  methods: [ GET, POST ]
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/config/routes.yaml:620]: TODO:: remove

File: src/plugins/orangehrmCorePlugin/Service/ConfigService.php [hunk 7]
```diff
@@ -501,14 +501,18 @@ public function getTimesheetTimeFormatConfig(): ?string
      */
     public function getLDAPSetting(): LDAPSetting
     {
-        return LDAPSetting::fromString($this->_getConfigValue(self::KEY_LDAP_SETTINGS));
+        $ldapSetting = $this->_getConfigValue(self::KEY_LDAP_SETTINGS);
+        if (is_null($ldapSetting)) {
+            return new LDAPSetting('localhost', 389, 'OpenLDAP', 'none', 'dc=example,dc=com');
```
Comment (RajithaKumara) [inline src/plugins/orangehrmCorePlugin/Service/ConfigService.php:506]:
```suggestion
            return new LDAPSetting('localhost', 389, 'OpenLDAP', 'none', '');
```
shall we set this empty :thinking:

File: src/plugins/orangehrmLDAPAuthenticationPlugin/Api/LDAPConfigAPI.php [hunk 8]
```diff
@@ -77,36 +79,23 @@ class LDAPConfigAPI extends Endpoint implements CollectionEndpoint
     public const LDAP_IMPLEMENTATION_ACTIVE_DIRECTORY = 'ActiveDirectory';
```
Comment (RajithaKumara) [inline src/plugins/orangehrmLDAPAuthenticationPlugin/Api/LDAPConfigAPI.php:79]: Move this kind of const to `LDAPSetting`

File: src/plugins/orangehrmLDAPAuthenticationPlugin/Api/LDAPConfigAPI.php (hunk 2 above)
Comment (RajithaKumara) [reply src/plugins/orangehrmLDAPAuthenticationPlugin/Api/LDAPConfigAPI.php:148]:
```php
if (!$bindAnonymously) {
    // ....
} else {
    $ldapSetting->setBindUserDN(null);
    $ldapSetting->setBindUserPassword(null);
}
```
:thinking:

File: src/plugins/orangehrmLDAPAuthenticationPlugin/Dto/LDAPSetting.php [hunk 9]
```diff
@@ -24,31 +24,38 @@
...
+    private string $implementation = 'OpenLDAP';
     private string $version = '3';
     private bool $optReferrals = false;
 
     private bool $bindAnonymously = true;
     private ?string $bindUserDN = null;
     private ?string $bindUserPassword = null;
 
-    private ?string $baseDN = null;
+    private ?string $baseDN = 'dc=example,dc=com';
```
Comment (RajithaKumara) [inline src/plugins/orangehrmLDAPAuthenticationPlugin/Dto/LDAPSetting.php:39]:
```suggestion
    private ?string $baseDN = '';
```

**Review by RajithaKumara — APPROVED** [review]

File: src/plugins/orangehrmLDAPAuthenticationPlugin/Dto/LDAPSetting.php (hunk 9 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmLDAPAuthenticationPlugin/Dto/LDAPSetting.php:39]: @Super-Chama default value for Base Distingushed Name

File: installer/Migration/V5_2_0/lang-string/admin.yaml [hunk 10]
```diff
@@ -71,3 +71,126 @@ langStrings:
...
+    value: 'Data Mapping'
+    unitId: data_mapping
+  -
+    value: 'Additional Settings'
+    unitId: additional_settings
+  -
+    value: 'Host'
+    unitId: host
+  -
+    value: 'LDAP Server IP or Hostname without the protocol (without ldp:// or ldps://)'
```
Comment (RajithaKumara) [inline installer/Migration/V5_2_0/lang-string/admin.yaml:93]:
```suggestion
    value: 'LDAP Server IP or Hostname without the protocol (without ldap:// or ldaps://)'
```

Comment (Super-Chama) [reply installer/Migration/V5_2_0/lang-string/admin.yaml:93]: Fixed https://github.com/RajithaKumara/orangehrm/pull/48

**Pull request merged**